### PR TITLE
Use new TLS key to sign DB cert during rotation

### PR DESCRIPTION
**Pull request by jakule**
Why
---
Current CA rotation require you to rotate the key 3 times
* `init` - to add new CA to the DB
* `update_clients` - to generate new client certificate and add it to DB 
* `standby` to remove the old CA from DB.

If any of the steps is skipped the integration will stop working as either Teleport or DB are not able to authenticate the connection.

Solution
---
Signing the client certificate with the new CA will allow to reduce the number of certificate updates and potencial DB restarts/reloads that should make the procedure simpler and less error prone. From now on the certificate will only need to be updated on `init` phase and after `standby` to remove the old CA.

### CODE REVIEW

File: lib/auth/db.go [hunk 1]
```diff
@@ -88,6 +90,17 @@ func (s *Server) GenerateDatabaseCert(ctx context.Context, req *proto.DatabaseCe
 	}, nil
 }
 
+// getCaAndSigner returns correct signer and CA that should be used when generating database certificate.
+// This function covers the database CA rotation scenario when on rotation init phase additional/new TLS
+// key should be used to sign the database CA. Otherwise, the trust chain will break on update_clients.
+func getCaAndSigner(keyStore keystore.KeyStore, databaseCA types.CertAuthority, req *proto.DatabaseCertRequest) ([]byte, crypto.Signer, error) {
```
Comment (jimbishopp) [inline lib/auth/db.go:96]:
```suggestion
func getCAandSigner(keyStore keystore.KeyStore, databaseCA types.CertAuthority, req *proto.DatabaseCertRequest) ([]byte, crypto.Signer, error) {
```
```suggestion
func getCaAndSigner(keyStore keystore.KeyStore, databaseCA types.CertAuthority, req *proto.DatabaseCertRequest) ([]byte, crypto.Signer, error) {
```

Comment (jimbishopp) [inline lib/auth/db.go:96]: Names like this are always hard to make readable. If you inline it you don't need a name :)

```suggestion
func getCAandSigner(keyStore keystore.KeyStore, databaseCA types.CertAuthority, req *proto.DatabaseCertRequest) ([]byte, crypto.Signer, error) {
```

**Review by jimbishopp — COMMENTED** [review]
Looks good. One minor comment. I hadn't reviewed this yet because of the conflict.

File: api/client/proto/authservice.proto [hunk 2]
```diff
@@ -788,6 +788,13 @@ message DatabaseCertRequest {
     int64 TTL = 3 [ (gogoproto.jsontag) = "ttl", (gogoproto.casttype) = "Duration" ];
     // ServerNames are SANs to include in the certificate.
     repeated string ServerNames = 4 [ (gogoproto.jsontag) = "server_names" ];
+    // Sender is a name of service that sent the request.
+    enum Sender {
+        UNSPECIFIED = 0;
+        TCTL = 1;
+    }
+    // Requester is sender of the message.
+    Sender Requester = 5 [ (gogoproto.jsontag) = "requester" ];
```
Comment (greedy52) [inline api/client/proto/authservice.proto:797]: nit: I think I would prefer to use either `Sender` or `Requester`, but not two different names at the same time. just personal preference, you can keep if you like.

File: lib/auth/db.go [hunk 3]
```diff
@@ -88,6 +90,17 @@ func (s *Server) GenerateDatabaseCert(ctx context.Context, req *proto.DatabaseCe
 	}, nil
 }
 
+// getCaAndSigner returns correct signer and CA that should be used when generating database certificate.
```
Comment (greedy52) [inline lib/auth/db.go:93]: `getCAAndSigner`? (comment and function name don't match)

**Review by greedy52 — APPROVED** [review]

File: lib/auth/db.go (hunk 3 above)
Comment (jakule) [reply lib/auth/db.go:93]: Thanks, fixed.

**Review by smallinsky — APPROVED** [review]
Sees reasonable. I have left some minor comments but other than that it LGTM.

@espadolini Could you please also take a look.

**Comment by espadolini** [comment]
What's meant to happen if the rotation is rolled back? Wouldn't you have the exact same problem?

**Comment by espadolini** [comment]
What's meant to happen if the rotation is rolled back? Wouldn't you have the exact same problem?

**Comment by jakule** [comment]
> What's meant to happen if the rotation is rolled back? Wouldn't you have the exact same problem?

@espadolini Yes and no. Let me explain.
Current CA rotation require you to rotate the key 3 times, `init` - to add new CA to the DB, `update_clients` - to generate new client certificate and add it to DB, `standby` to remove the old CA from DB. The last step is "recommended", as if skipped it won't break the integration. The old CA should be removed to reduce the surface of potencial attack.
After this change the second rotation on `update_clients` can be skipped as the client cert will be already signed by the new cert. 
Now answering your question. In case of a rollback, the certificate needs to be updated on the DB side 3 times, including the one on rollback as we need the old client certificate back. Any changes are not needed, as `rollback` already uses the "old" CA to sign client certificate.

**Review by espadolini — APPROVED** [review]
@jakule ty for the explanation, it makes more sense as a convenience measure just for `tctl`. I'm not sure I like the fact that database certs end up being handled differently than other types of certs in `tctl auth`, but they have separate codepaths anyway.

**Pull request merged**